transports/webrtc-websys/src/upgrade.rs

```diff
@@ -0,0 +1,152 @@
+pub(crate) mod noise;
+
+pub(crate) use super::fingerprint::Fingerprint;
+use super::stream::{DataChannelConfig, WebRTCStream};
+use super::utils;
+use super::Error;
+use super::{sdp, Connection};
+use js_sys::{Object, Reflect};
+use libp2p_identity::{Keypair, PeerId};
+use log::debug;
+use send_wrapper::SendWrapper;
+use std::net::SocketAddr;
+use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::JsFuture;
+use web_sys::{MessageEvent, RtcConfiguration, RtcDataChannel, RtcPeerConnection};
+
+const SHA2_256: u64 = 0x12;
+const SHA2_512: u64 = 0x13;
+
+pub struct Upgrader {
+    inner: SendWrapper<UpgraderInner>,
+}
+
+impl Upgrader {
+    pub fn new() -> Self {
+        Self {
+            inner: SendWrapper::new(UpgraderInner::new()),
+        }
+    }
+
+    pub(crate) async fn outbound(
+        &mut self,
+        sock_addr: SocketAddr,
+        remote_fingerprint: Fingerprint,
+        id_keys: Keypair,
+    ) -> Result<(PeerId, Connection), Error> {
+        let fut = SendWrapper::new(self.inner.outbound(sock_addr, remote_fingerprint, id_keys));
+        fut.await
+    }
+}
+
+pub struct UpgraderInner;
```

Comment:
Why do we need this struct if it doesn't carry any state? Can't we just have two top-level free functions?

Reply:
Good point! We do not need it, I'm reducing it down to two top-level free functions.